File: www/shared-memory/sysV_out.php
// sysV_out.php
<?php 
################################################################################
# System-V IPC for PHP - Semaphores, Shared Memory and Message Queues
# https://www.php.net/manual/en/book.shmop.php
# https://www.php.net/manual/en/book.sem.php
################################################################################
date_default_timezone_set('Europe/Berlin'); # set the default timezone to use

# create any file to generate the specific FTOK key
$filePathName = '/tmp/example';   # def a filepath / filename
$fd = fopen($filePathName, "w");  # create file-handle for writing
fclose($fd);                      # close file-handle

# generate the specific FTOK key from file-path-name
#  ftok(string $filename, string $project_id): int
#      "id" must be as chr(string) - hex(0x30) = dec(48) = chr("0") 
$ftokKey = ftok($filePathName, "0");
print("ftokKey=" . $ftokKey . " hex=0x" . dechex($ftokKey) . PHP_EOL);

while (1) {
    $TSnow = date('d.m.YTH:i:s');        # create TimeStamp
    $sem_id = @sem_get($ftokKey);        # creating a semaphore
    if (False == $sem_id) {
        print ("can't create or open the Semaphore with KEY: " . $ftokKey . PHP_EOL);
        print ("Please check Semaphore permission!" . PHP_EOL);
        break;
    }

    if (sem_acquire($sem_id)) {          # acquiring the semaphore
                                         # create shared memeory object
        $shdMemObj = shmop_open($ftokKey, "c", 0644, strlen($TSnow));
        if (!$shdMemObj) {
            echo "Couldn't create shared memory segment\n";
        }

        $shm_bytes_written = shmop_write($shdMemObj, $TSnow, 0);
        shmop_close($shdMemObj);

        if ($shm_bytes_written == strlen($TSnow))
             print ("  len=" . $shm_bytes_written . "  TSnow=" . $TSnow . PHP_EOL);
        else print ("  Couldn't write the entire length of data" . PHP_EOL);

        sem_release($sem_id);            # release the semaphore
        sleep(1);
    }
}

# sem_remove($sem_id);                     # delete a semaphore

echo (PHP_EOL . "To control shared-memory and semaphore, please call 'ipcs' or remove with 'ipcrm -r ID'" . PHP_EOL);
echo ("look at: 'cat /proc/sysvipc/shm'" . PHP_EOL);
?>

File: www/shared-memory/sysV_out.py
# sysV_out.py
################################################################################
# System V IPC for Python - Semaphores, Shared Memory and Message Queues
# https://semanchuk.com/philip/sysv_ipc/
################################################################################

import os, time
from datetime import datetime
import sysv_ipc

# create any file to generate the specific FTOK key
filePathName = '/tmp/example'                  # def a filepath / filename
fd = os.open(filePathName, flags=os.O_CREAT)   # create file-handle
os.close(fd)                                   # close file-handle

# generate the specific FTOK key from file-path-name
# ftok(path, id, [silence_warning = False])
# Note that ftok() has limitations, and this function will issue a warning 
# to that effect unless silence_warning is True. 
# "id" must be in hex - hex(0x30) = dec(48) = chr("0") 
ftokKey = sysv_ipc.ftok(filePathName, 0x30, silence_warning = True)
print(f"{ftokKey=} hex=0x{ftokKey:08x}")

while True:
    # create TimeStamp
    TSnow = datetime.now().strftime('%d.%m.%YT%H:%M:%S')

    # create shared memeory object
    # PC_CREAT used when creating IPC objects. 
    shdMemObj = sysv_ipc.SharedMemory(ftokKey, size=len(TSnow), flags=sysv_ipc.IPC_CREAT, mode=0o644)
    shdMemObj.write(TSnow)
    time.sleep(1)

    print(f" {shdMemObj.id=} len={len(TSnow)} {TSnow=}")
    shdMemObj.detach()
# shdMemObj.remove()

print(f"check Shared Memory Segments with command 'ipcs -m'")
print(f"and look at: 'cat /proc/sysvipc/shm'")
